fmpz_poly/div_series_basecase.c

```diff
@@ -31,59 +43,147 @@ _fmpz_poly_div_series_basecase(fmpz * Q, const fmpz * A, slong Alen,
-            }
+        /* We have small coefficients for i < nsmall */
+        for (nsmall = 0; nsmall < Blen; nsmall++)
+        {
+            b = B[nsmall];
+
+            if (COEFF_IS_MPZ(b))
+                break;
+
+            b = FLINT_ABS(b);
+            if (nsmall == 0 || (b >> Bbits[nsmall - 1]))
+                Bbits[nsmall] = FLINT_BIT_COUNT(b);
+            else
+                Bbits[nsmall] = Bbits[nsmall - 1];
         }
 
         for (i = 1; i < n; i++)
         {
-            fmpz_mul(Q + i, B + 1, Q + i - 1);
+            if (i >= nsmall || Qbits > FLINT_BITS - 2 || Bbits[i] > FLINT_BITS - 2)
+            {
+                /* Can't use fast code. */
+                bits = WORD_MAX;
+            }
+            else
+            {
+                /* Can maybe use fast code; bound bits. */
+                b = FLINT_MIN(i, Blen - 1);
+                bits = FLINT_BIT_COUNT(b);
+
+                /* Bit size of product. */
+                bits += Bbits[i] + Qbits;
+
+                /* Sign. */
+                bits += 1;
+            }
+
+            if (bits <= 3 * FLINT_BITS - 1)
+            {
+                if (bits <= FLINT_BITS - 1)
+                {
+                    slong s, x, y;
+
+                    s = 0;
+
+                    for (j = 1; j < FLINT_MIN(i + 1, Blen); j++)
+                    {
+                        x = B[j];
+                        y = Q[i - j];
+                        s += x * y;
+                    }
 
-            for (j = 2; j < FLINT_MIN(i + 1, Blen); j++)
-                fmpz_addmul(Q + i, B + j, Q + i - j);
+                    fmpz_set_si(Q + i, s);
+                }
+                else if (bits <= 2 * FLINT_BITS - 1)
+                {
+                    mp_limb_t hi, lo, shi, slo;
+                    slong x, y;
+
+                    shi = slo = 0;
+
+                    for (j = 1; j < FLINT_MIN(i + 1, Blen); j++)
+                    {
+                        x = B[j];
+                        y = Q[i - j];
+
+                        smul_ppmm(hi, lo, x, y);
+                        add_ssaaaa(shi, slo, shi, slo, hi, lo);
+                    }
+
+                    fmpz_set_signed_uiui(Q + i, shi, slo);
+                }
+                else
+                {
+                    mp_limb_t hi, lo, cy, shh, shi, slo;
+                    slong x, y;
+
+                    shh = shi = slo = 0;
+
+                    for (j = 1; j < FLINT_MIN(i + 1, Blen); j++)
+                    {
+                        x = B[j];
+                        y = Q[i - j];
+
+                        smul_ppmm(hi, lo, x, y);
+                        add_sssaaaaaa(cy, shi, slo, 0, shi, slo, 0, hi, lo);
+                        shh += (0 <= (slong) hi) ? cy : cy - 1;
+                    }
+
+                    fmpz_set_signed_uiuiui(Q + i, shh, shi, slo);
+                }
+            }
+            else
+            {
+                fmpz_mul(Q + i, B + 1, Q + i - 1);
```

Comment:
This looks unprotected. What if B has length 1?

Reply:
Blen == 1 is handled separately at the start of the function.

Reply:
Ok, I didn't check that. I only saw the Alen == 1 case being handled.